Separate h and p refinement criteria so they can be applied separately in evolution executables

kidder opened the pull request:
## Proposed changes

- introduce enum amr::Criteria::Type to designate each amr::Criterion as being h or p
- only evaluate p criteria during the event RefineMesh
- add bool p_refine_only_in_event to amr struct in metavariables that allows one to choose whether or not to evaluate p criteria during AMR phase changes

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
- New amr::Criterion have to overload the function `type` based on whether it returns h or p refinement flags
- The amr struct in the metavariables must define a static constexpr bool p_refine_in_events_only; the recommendation is to set it to true for evolution executables and false for elliptic executables.
- If you were previously doing p-refinement via phase changes, you will need to add something like the following to the list of EventsAndTriggers in the input file:
```
 - Trigger:
       Slabs:
       EvenlySpaced:
           Interval: 1
           Offset: 0
    Events:
        RefineMesh
```
 where you should change the trigger to what you desire.   If you only want to do p-refinement, then you should omit PhaseChanges related to AMR in the list of PhaseChangeAndTriggers.

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder committed:
Add enum for type of AMR criteria

kidder committed:
Template DriveToTarget on amr::Criteria::Type

This allows separate criteria for h and p refinement

kidder committed:
Add virtual function to amr::Criterion for its Type

This will allow h and p refinement to be done separately.

kidder committed:
Only evaluate p-refinement criteria in RefineMesh

Previously h-refinement criteria were evaluated, but then ignored.
Now they will be skipped.

kidder committed:
Add constexpr bool p_refine_only_in_event to AmrMetavariables

If set to true, this bool will only allow h-refinement during the
AMR phases.  p-refinement should then be done via
amr::Event::RefineMesh

nilsvu commented on tests/Unit/Elliptic/DiscontinuousGalerkin/SubdomainOperator/Test_SubdomainOperator.cpp:
```diff
@@ -607,11 +609,8 @@ void test_subdomain_operator(
-            {::amr::Flag::IncreaseResolution, 1},
-            {::amr::Flag::DoNothing, 1}}));
+    amr_criteria.push_back(
+        std::make_unique<::amr::Criteria::IncreaseResolution<Dim>>());
```
Hm the idea here is to create random p-nonconforming boundaries and make sure the subdomain operator handles them all correctly. Just uniformly increasing p-refinement is a much weaker test. I don't immediately have a good solution.

kidder replied:
I should be able to template Random on the refinement type just like DriveToTarget if you want the stronger test...

nilsvu commented on src/ParallelAlgorithms/Amr/Actions/EvaluateRefinementCriteria.hpp:
```diff
@@ -56,7 +57,8 @@ struct get_tags {
 /// \brief Evaluates the refinement criteria in order to set the amr::Info of an
-/// Element and sends this information to the neighbors of the Element.
+/// Element and sends this information to the neighbors of the Element.  If
+/// DoNotRefineMesh is true, only do h-refinement.
```
update docs to refer to `Metavariables::amr::p_refine_only_in_event`

wthrowe reviewed (dismissed):
I suggest including something in the upgrade instructions on how to fix input files for anyone doing evolution p-AMR via a phase.

Edit: @kidder or any other core dev: feel free to merge this once the upgrade instructions are done.  There are known users whose runs this will break, so I don't want to merge without them.

kidder commented:
I've updated the upgrade instructions.

kidder committed:
Template Random on amr::Criteria::Type

This allows separate criteria for h and p refinement.

kidder committed:
Trigger both h and p refinement in 1D ScalarWave

kidder commented:
@nilsvu @wthrowe I've updated this PR:
- templated Random on refinement type
- made the minor doc change requested
- updated the 1D SW input files to do both h and p refinement.

wthrowe commented:
Looks fine, but CI says you have whitespace issues.
